consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/statemachine/BaseBftController.java

```diff
@@ -0,0 +1,210 @@
+import org.hyperledger.besu.consensus.common.bft.Gossiper;
+import org.hyperledger.besu.consensus.common.bft.MessageTracker;
+import org.hyperledger.besu.consensus.common.bft.SynchronizerUpdater;
+import org.hyperledger.besu.consensus.common.bft.events.BftReceivedMessageEvent;
+import org.hyperledger.besu.consensus.common.bft.events.BlockTimerExpiry;
+import org.hyperledger.besu.consensus.common.bft.events.NewChainHead;
+import org.hyperledger.besu.consensus.common.bft.events.RoundExpiry;
+import org.hyperledger.besu.consensus.common.bft.messagewrappers.BftMessage;
+import org.hyperledger.besu.consensus.common.bft.payload.Authored;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.Message;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.MessageData;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public abstract class BaseBftController implements BftEventHandler {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private final Blockchain blockchain;
+  private final BftFinalState bftFinalState;
+  private final FutureMessageBuffer futureMessageBuffer;
+  private final Gossiper gossiper;
+  private final MessageTracker duplicateMessageTracker;
+  private final SynchronizerUpdater sychronizerUpdater;
+
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  protected BaseBftController(
+      final Blockchain blockchain,
+      final BftFinalState bftFinalState,
+      final Gossiper gossiper,
+      final MessageTracker duplicateMessageTracker,
+      final FutureMessageBuffer futureMessageBuffer,
+      final SynchronizerUpdater sychronizerUpdater) {
+    this.blockchain = blockchain;
+    this.bftFinalState = bftFinalState;
+    this.futureMessageBuffer = futureMessageBuffer;
+    this.gossiper = gossiper;
+    this.duplicateMessageTracker = duplicateMessageTracker;
+    this.sychronizerUpdater = sychronizerUpdater;
+  }
+
+  @Override
+  public void start() {
+    if (started.compareAndSet(false, true)) {
+      startNewHeightManager(blockchain.getChainHeadHeader());
+    }
+  }
+
+  @Override
+  public void handleMessageEvent(final BftReceivedMessageEvent msg) {
+    final MessageData data = msg.getMessage().getData();
+    if (!duplicateMessageTracker.hasSeenMessage(data)) {
+      duplicateMessageTracker.addSeenMessage(data);
+      handleMessage(msg.getMessage());
+    } else {
+      LOG.trace("Discarded duplicate message");
+    }
+  }
+
+  protected abstract void handleMessage(final Message message);
+
+  protected <P extends BftMessage<?>> void consumeMessage(
+      final Message message, final P ibftMessage, final Consumer<P> handleMessage) {
+    LOG.trace("Received IBFT {} message", ibftMessage.getClass().getSimpleName());
```

Comment:
nit: trace log refers to IBFT. As well as method arg name is ibftMessage.

Reply:
good catch.